### PR TITLE
Add is_embedded attribute on ObjectSchema

### PR DESCRIPTION
**Pull request by jedelbo**

### CODE REVIEW

**Comment by jedelbo** [comment]
@tgoyne @cmelchior I am not sure what kind of support for embedded object is expected here. Is this PR a step in the right direction?

**Comment by tgoyne** [comment]
Embeddedness is a property of ObjectSchema, not Property.

**Comment by kneth** [comment]
Will it also validate the schema? For example, an embedded object cannot define a primary key.

File: src/object_schema.cpp [hunk 1]
```diff
@@ -330,6 +343,9 @@ void ObjectSchema::validate(Schema const& schema, std::vector<ObjectSchemaValida
         validate_property(schema, name, prop, &primary, exceptions);
     }
 
+    if (!primary_key.empty() && is_embedded) {
+        exceptions.emplace_back("Embedded table '%1' cannot have primary key '%2'.", name, primary_key);
```
Comment (cmelchior) [inline src/object_schema.cpp]: Just to make it more clear that no primary key is allowed at all.
```suggestion
        exceptions.emplace_back("Embedded table '%1' is not allowed to define a primary key. Found key '%2'.", name, primary_key);
```

File: tests/schema.cpp [hunk 2]
```diff
@@ -450,6 +454,16 @@ TEST_CASE("Schema") {
             REQUIRE_THROWS_CONTAINING(schema.validate(), "Properties 'pk2' and 'pk1' are both marked as the primary key of 'object'.");
         }
 
+        SECTION("rejects primary key on embedded table") {
```
Comment (cmelchior) [inline tests/schema.cpp]: We probably also need a migration test, e.g. if a local Realm adds or remove the embeddedness of a table during a migration.

File: tests/realm.cpp [hunk 3]
```diff
@@ -1284,6 +1284,25 @@ TEST_CASE("SharedRealm: dynamic schema mode doesn't invalidate object schema poi
     REQUIRE(object_schema == &*r2->schema().find("object"));
 }
 
+TEST_CASE("SharedRealm: declaring an object as embedded results in creating an embedded table") {
```
Comment (tgoyne) [inline tests/realm.cpp]: The existing tests for validating that we created the correct things in core from a given schema are in migrations.cpp. Adding sections for is_embedded to each of the existing test cases there will probably reveal some more scenarios that need to be handled.

File: tests/schema.cpp (hunk 2 above)
Comment (jedelbo) [reply tests/schema.cpp]: It is not possible to change embeddedness of a table after creation

Comment (cmelchior) [reply tests/schema.cpp]: May I ask why not? For local Realms it just seems to be a boolean and some rules for transactions?

Comment (jedelbo) [reply tests/schema.cpp]: Yes, it is just a flag, but the flag indicates that some invariants is upheld. We could probably check if a non embedded table complies to these invariants before making it embedded. This requirement was not captured in the tecnical design.

Comment (finnschiermer) [reply tests/schema.cpp]: Actually - the first implementation allowed this. And then we simplified. Apparently too much.

File: tests/realm.cpp (hunk 3 above)
Comment (jedelbo) [reply tests/realm.cpp]: Ok, I will look into that.

**Comment by jedelbo** [comment]
@tgoyne Could you have a look at the last commit where I try to create embedded objects based on values in a context? Is it ok to do it that way?

**Comment by jedelbo** [comment]
I will add tests around changing embeddedness in a separate PR

**Pull request merged**